### PR TITLE
provider/docker: Trigger listener

### PR DESCRIPTION
**Pull request by lwander**
Needs https://github.com/spinnaker/igor/pull/65

With this change, and a custom pipeline sent to front50, I was able to trigger a pipeline off a push to github registered to a docker repository on dockerhub.

### CODE REVIEW

**Comment by lwander** [comment]
@duftler

File: echo-pipelinetriggers/src/main/java/com/netflix/spinnaker/echo/pipelinetriggers/monitor/TriggerMonitor.java [hunk 1]
```diff
@@ -41,68 +41,70 @@
...
-
-    protected void onEventProcessed(final TriggerEvent event) {
-        registry.counter("echo.events.processed").increment();
-    }
-
-    private void onSubscriberError(Throwable error) {
-        log.error("Subscriber raised an error processing pipeline", error);
-        registry.counter("trigger.errors").increment();
-    }
+  protected final Action1<Pipeline> subscriber;
```
Comment (lwander) [inline echo-pipelinetriggers/src/main/java/com/netflix/spinnaker/echo/pipelinetriggers/monitor/TriggerMonitor.java]: Here I realized my last PR added this file with bad identation.

File: echo-model/src/main/groovy/com/netflix/spinnaker/echo/model/trigger/DockerEvent.groovy [hunk 2]
```diff
@@ -16,5 +16,23 @@
 
 package com.netflix.spinnaker.echo.model.trigger
 
+import com.fasterxml.jackson.annotation.JsonIgnoreProperties
+import groovy.transform.Canonical
+
+@Canonical
+@JsonIgnoreProperties(ignoreUnknown = true)
 class DockerEvent extends TriggerEvent {
+  static final public String TYPE = "DOCKER"
```
Comment (duftler) [inline echo-model/src/main/groovy/com/netflix/spinnaker/echo/model/trigger/DockerEvent.groovy]: Can drop `public` here and the semi-colons down below.

File: echo-model/src/main/groovy/com/netflix/spinnaker/echo/model/trigger/DockerEvent.groovy [hunk 3]
```diff
@@ -16,5 +16,23 @@
...
+
+  Content content;
+
+  @Canonical
+  @JsonIgnoreProperties(ignoreUnknown = true)
+  public static class Content {
```
Comment (duftler) [inline echo-model/src/main/groovy/com/netflix/spinnaker/echo/model/trigger/DockerEvent.groovy]: Can drop `public` here.

File: echo-model/src/main/groovy/com/netflix/spinnaker/echo/model/trigger/GitEvent.groovy [hunk 4]
```diff
@@ -23,22 +23,22 @@ import groovy.transform.Canonical
 @Canonical
 @JsonIgnoreProperties(ignoreUnknown = true)
 class GitEvent extends TriggerEvent {
-    Content content;
+  Content content;
```
Comment (duftler) [inline echo-model/src/main/groovy/com/netflix/spinnaker/echo/model/trigger/GitEvent.groovy]: Can drop `public`'s and semi-colons.

File: echo-pipelinetriggers/src/main/java/com/netflix/spinnaker/echo/pipelinetriggers/monitor/DockerEventMonitor.java [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+
+  public static final String TRIGGER_TYPE = "docker";
+
+  private final ObjectMapper objectMapper = new ObjectMapper();
+
+  private final PipelineCache pipelineCache;
+
+  @Autowired
+  public DockerEventMonitor(@NonNull PipelineCache pipelineCache,
+                @NonNull Action1<Pipeline> subscriber,
```
Comment (duftler) [inline echo-pipelinetriggers/src/main/java/com/netflix/spinnaker/echo/pipelinetriggers/monitor/DockerEventMonitor.java]: Indentation is off for method parameters.

File: echo-pipelinetriggers/src/main/java/com/netflix/spinnaker/echo/pipelinetriggers/monitor/DockerEventMonitor.java [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+                @NonNull Registry registry) {
+    super(subscriber, registry);
+    this.pipelineCache = pipelineCache;
+  }
+
+  @Override
+  public void processEvent(Event event) {
+    if (!event.getDetails().getType().equalsIgnoreCase(DockerEvent.TYPE)) {
```
Comment (duftler) [inline echo-pipelinetriggers/src/main/java/com/netflix/spinnaker/echo/pipelinetriggers/monitor/DockerEventMonitor.java]: Can `getDetails()` return null here?

File: echo-pipelinetriggers/src/main/java/com/netflix/spinnaker/echo/pipelinetriggers/monitor/DockerEventMonitor.java [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+      );
+  }
+
+  @Override
+  protected Predicate<Trigger> matchTriggerFor(final TriggerEvent event) {
+    DockerEvent dockerEvent = (DockerEvent) event;
+    String registry = dockerEvent.getContent().getRegistry();
+    String repository = dockerEvent.getContent().getRepository();
+    String tag = dockerEvent.getContent().getTag();
+    return trigger ->  trigger.getType().equals(TRIGGER_TYPE) &&
```
Comment (duftler) [inline echo-pipelinetriggers/src/main/java/com/netflix/spinnaker/echo/pipelinetriggers/monitor/DockerEventMonitor.java]: Not sure what the typical style is here, but there are two spaces after the `->` operator.

File: echo-pipelinetriggers/src/main/java/com/netflix/spinnaker/echo/pipelinetriggers/monitor/TriggerMonitor.java [hunk 8]
```diff
@@ -41,68 +41,70 @@
...
+  protected final Registry registry;
+
+  public TriggerMonitor(@NonNull Action1<Pipeline> subscriber,
+              @NonNull Registry registry) {
```
Comment (duftler) [inline echo-pipelinetriggers/src/main/java/com/netflix/spinnaker/echo/pipelinetriggers/monitor/TriggerMonitor.java]: Indentation is off for method parameters.

File: echo-pipelinetriggers/src/test/groovy/com/netflix/spinnaker/echo/pipelinetriggers/DockerEventMonitorSpec.groovy [hunk 9]
```diff
@@ -0,0 +1,172 @@
+/*
+ * Copyright 2016 Netflix, Inc.
```
Comment (duftler) [inline echo-pipelinetriggers/src/test/groovy/com/netflix/spinnaker/echo/pipelinetriggers/DockerEventMonitorSpec.groovy]: Copyright is off.

File: echo-pipelinetriggers/src/test/groovy/com/netflix/spinnaker/echo/pipelinetriggers/DockerEventMonitorSpec.groovy [hunk 10]
```diff
@@ -0,0 +1,172 @@
...
+    pipelineCache.getPipelines() >> [pipeline]
+
+    when:
+    monitor.processEvent(objectMapper.convertValue(event, Event))
+
+    then:
+    0 * subscriber._
+
+    where:
+    trigger                                 | description
```
Comment (duftler) [inline echo-pipelinetriggers/src/test/groovy/com/netflix/spinnaker/echo/pipelinetriggers/DockerEventMonitorSpec.groovy]: Could remove some of the whitespace before the `|`.

File: echo-pipelinetriggers/src/test/groovy/com/netflix/spinnaker/echo/pipelinetriggers/DockerEventMonitorSpec.groovy [hunk 11]
```diff
@@ -0,0 +1,172 @@
...
+    monitor.processEvent(objectMapper.convertValue(event, Event))
+
+    then:
+    0 * subscriber._
+
+    where:
+    trigger                                               | description
+    disabledDockerTrigger                                 | "disabled docker trigger"
+    enabledDockerTrigger.withRegistry("notRegistry")      | "different registry"
+    enabledDockerTrigger.withRepository("notRepository") | "different repository"
```
Comment (duftler) [inline echo-pipelinetriggers/src/test/groovy/com/netflix/spinnaker/echo/pipelinetriggers/DockerEventMonitorSpec.groovy]: Pipe alignment is off here.

**Comment by duftler** [comment]
LGTM after addressing comments.

File: echo-pipelinetriggers/src/test/groovy/com/netflix/spinnaker/echo/pipelinetriggers/DockerEventMonitorSpec.groovy [hunk 12]
```diff
@@ -0,0 +1,172 @@
...
+    pipeline = createPipelineWith(trigger)
+    event = createDockerEvent()
+  }
+
+
+  @Unroll
+  def "does not trigger a pipeline that has an enabled docker trigger with missing #field"() {
+    given:
+    pipelineCache.getPipelines() >> [badPipeline, goodPipeline]
+    println objectMapper.writeValueAsString(createDockerEvent())
```
Comment (duftler) [inline echo-pipelinetriggers/src/test/groovy/com/netflix/spinnaker/echo/pipelinetriggers/DockerEventMonitorSpec.groovy]: Extraneous `println`.

File: echo-pipelinetriggers/src/main/java/com/netflix/spinnaker/echo/pipelinetriggers/monitor/DockerEventMonitor.java (hunk 6 above)
Comment (lwander) [inline echo-pipelinetriggers/src/main/java/com/netflix/spinnaker/echo/pipelinetriggers/monitor/DockerEventMonitor.java]: I'd rather throw an NPE here than silently ignore invalid events being passed through echo. WDYT?

Comment (lwander) [inline echo-pipelinetriggers/src/main/java/com/netflix/spinnaker/echo/pipelinetriggers/monitor/DockerEventMonitor.java]: ping @duftler (I addressed the other comments).

File: echo-model/src/main/groovy/com/netflix/spinnaker/echo/model/trigger/DockerEvent.groovy (hunk 2 above)
Comment (lwander) [inline echo-model/src/main/groovy/com/netflix/spinnaker/echo/model/trigger/DockerEvent.groovy]: You can only drop `public` for methods.

File: echo-pipelinetriggers/src/main/java/com/netflix/spinnaker/echo/pipelinetriggers/monitor/DockerEventMonitor.java (hunk 6 above)
Comment (duftler) [inline echo-pipelinetriggers/src/main/java/com/netflix/spinnaker/echo/pipelinetriggers/monitor/DockerEventMonitor.java]: I would prefer to explicitly throw an exception that has enough detail to do some troubleshooting. I don't think the NPE message would contain details of the event itself.

Comment (lwander) [inline echo-pipelinetriggers/src/main/java/com/netflix/spinnaker/echo/pipelinetriggers/monitor/DockerEventMonitor.java]: Validation is now done in the super class.

**Pull request merged**